atlasdb-commons/src/test/java/com/palantir/async/initializer/LambdaCallbackTest.java

```diff
@@ -0,0 +1,132 @@
+
+        assertThatThrownBy(() -> callback.runWithRetry(initsAndCleanups)).isEqualToComparingFieldByField(INIT_FAIL);
+        assertThat(initsAndCleanups.inits()).isEqualTo(1);
+        assertThat(initsAndCleanups.cleanups()).isEqualTo(1);
+    }
+
+    @Test
+    public void singleAttemptCallbackThrowsCleanupExceptionOnCleanupFailure() {
+        callback = LambdaCallback.singleAttempt(this::markInitAndFail, this::markCleanupAndFail);
+
+        assertThatThrownBy(() -> callback.runWithRetry(initsAndCleanups)).isEqualToComparingFieldByField(CLEANUP_FAIL);
+    }
+
+    @Test
+    public void retryingCallbackCallsInitializeOnceAndSkipsCleanupOnSuccess() {
+        callback = LambdaCallback.retrying(this::markInit, this::markCleanup);
+
+        callback.runWithRetry(initsAndCleanups);
+        assertThat(initsAndCleanups.inits()).isEqualTo(1);
+        assertThat(initsAndCleanups.cleanups()).isEqualTo(0);
+    }
+
+    @Test
+    public void retryingCallbackCallsInitializeAndCleanupUntilSuccess() {
+        callback = LambdaCallback.retrying(this::markInitThenFailIfFewerThatTenInits, this::markCleanup);
+
+        callback.runWithRetry(initsAndCleanups);
+        assertThat(initsAndCleanups.inits()).isEqualTo(10);
+        assertThat(initsAndCleanups.cleanups()).isEqualTo(9);
+    }
+
+    @Test
+    public void retryingCallbackStopsRetryingWhenCleanupThrows() {
+        callback = LambdaCallback.retrying(this::markInitAndFail, this::markCleanupThenFailIfMoreThatTenInits);
+
+        assertThatThrownBy(() -> callback.runWithRetry(initsAndCleanups)).isEqualToComparingFieldByField(CLEANUP_FAIL);
+        assertThat(initsAndCleanups.inits()).isEqualTo(11);
+        assertThat(initsAndCleanups.cleanups()).isEqualTo(11);
+    }
+
+    private void markInit(ModifiableInitsAndCleanups initsAndCleanups) {
+        initsAndCleanups.setInits(initsAndCleanups.inits() + 1);
+    }
+
+    private void markInitAndFail(ModifiableInitsAndCleanups initsAndCleanups) {
+        markInit(initsAndCleanups);
+        throw INIT_FAIL;
+    }
+
+    private void markInitThenFailIfFewerThatTenInits(ModifiableInitsAndCleanups initsAndCleanups) {
+        markInit(initsAndCleanups);
+        if (initsAndCleanups.inits() < 10) {
+            throw INIT_FAIL;
+        }
+    }
+
+    private void markCleanup(ModifiableInitsAndCleanups initsAndCleanups, Throwable ignore) {
+        initsAndCleanups.setCleanups(initsAndCleanups.cleanups() + 1);
+    }
+
+    private void markCleanupAndFail(ModifiableInitsAndCleanups initsAndCleanups, Throwable ignore) {
+        markCleanup(initsAndCleanups, ignore);
+        throw CLEANUP_FAIL;
+    }
+
+    private void markCleanupThenFailIfMoreThatTenInits(ModifiableInitsAndCleanups initsAndCleanups, Throwable ignore) {
+        markCleanup(initsAndCleanups, ignore);
+        if (initsAndCleanups.inits() > 10) {
+            throw CLEANUP_FAIL;
+        }
+    }
+
+    @Value.Modifiable
```

Comment:
Interesting! Wasn't aware of this feature

Reply:
Yeah really useful, I saw it in one of the recent PR's